**Objectives:**
1. Check for missing values
2. How many Alice rejections total?
3. look at percentages of total applications compared to alice rejections in the class.
4. for 706 and 705: look at dataset for rows without class or art unit (fill in value based on available value)
5. look at percentages of total applications compared to alice rejections in the class.
6. create and export a csv file with only the 705 and 706 rejections

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as scs
import statsmodels.api as sm
import matplotlib.pyplot as plt

%matplotlib inline
%config InlineBackend.figure_format='retina'


# Load data/art_unit_alice/rejection_types_with_alice_2015_to_2017-07-11.csv

In [2]:
df=pd.read_csv('data/art_unit_alice/rejection_types_with_alice_2015_to_2017-07-11.csv', index_col=None)
print(df.columns)
print(df.info())
print(df.describe())
df.head().T

/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (2,5) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Index(['Unnamed: 0', 'app_id_x', 'ifw_number', 'document_cd', 'mail_dt',
       'art_unit', 'uspc_class', 'uspc_subclass', 'app_id_y', 'alice_in'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1778669 entries, 0 to 1778668
Data columns (total 10 columns):
Unnamed: 0       int64
app_id_x         int64
ifw_number       object
document_cd      object
mail_dt          object
art_unit         object
uspc_class       object
uspc_subclass    object
app_id_y         float64
alice_in         float64
dtypes: float64(2), int64(2), object(6)
memory usage: 135.7+ MB
None
         Unnamed: 0      app_id_x      app_id_y      alice_in
count  1.778669e+06  1.778669e+06  1.713149e+06  1.778669e+06
mean   1.989841e+06  1.408780e+07  1.408981e+07  2.581593e-02
std    1.191709e+06  5.774059e+05  5.832540e+05  1.585859e-01
min    0.000000e+00  1.200003e+07  1.200003e+07  0.000000e+00
25%    9.555980e+05  1.376726e+07  1.376612e+07  0.000000e+00
50%    1.991182e+06  1.413574e+07  1

0           1           2           3           4
Unnamed: 0              0           1           2           3           4
app_id_x         14150981    14198961    13796589    14673475    14669113
ifw_number      100867762   100867788   100867794   100867844   100867850
document_cd          CTFR        CTFR        CTNF        CTNF        CTNF
mail_dt        2015-10-15  2015-10-15  2015-10-15  2015-10-15  2015-10-15
art_unit             2632        2699        3776        2627        2666
uspc_class            375         345         606         345         382
uspc_subclass      219000      173000      159000      175000      118000
app_id_y              NaN         NaN         NaN         NaN         NaN
alice_in                0           0           0           0           0

## remove Unnamed column

In [3]:
df.drop('Unnamed: 0',axis=1, inplace=True)

In [4]:
df.head()

app_id_x ifw_number document_cd     mail_dt art_unit uspc_class  \
0  14150981  100867762        CTFR  2015-10-15     2632        375   
1  14198961  100867788        CTFR  2015-10-15     2699        345   
2  13796589  100867794        CTNF  2015-10-15     3776        606   
3  14673475  100867844        CTNF  2015-10-15     2627        345   
4  14669113  100867850        CTNF  2015-10-15     2666        382   

  uspc_subclass  app_id_y  alice_in  
0        219000       NaN       0.0  
1        173000       NaN       0.0  
2        159000       NaN       0.0  
3        175000       NaN       0.0  
4        118000       NaN       0.0

# 1. Check for missing values

In [29]:
print(df['art_unit'].isna().sum())
print(df['uspc_class'].isna().sum())
print(df['uspc_subclass'].isna().sum())

0
0
0


In [30]:
print(df['art_unit'].isnull().sum())
print(df['uspc_class'].isnull().sum())
print(df['uspc_subclass'].isnull().sum())

0
0
0


In [32]:
count_nan = len(df) - df.count()
count_nan

app_id_x             0
ifw_number           0
document_cd          0
mail_dt              0
art_unit             0
uspc_class           0
uspc_subclass        0
app_id_y         65520
alice_in             0
dtype: int64

# 2. How many alice rejections total?
**45,918** alice rejections out of 1,778,669 applications with office actions.

In [5]:
df['alice_in'].value_counts()

0.0    1732751
1.0      45918
Name: alice_in, dtype: int64

# issue where I get different value counts

In [34]:
df[df['art_unit']=='0']
#17139

app_id_x       ifw_number document_cd     mail_dt art_unit  \
80951    14091028        103154643        CTFR  2015-12-23        0   
80964    14517473        103154799        CTNF  2015-12-23        0   
80969    14033204        103154894        CTFR  2015-12-23        0   
80974    14034169        103156096        CTFR  2015-12-23        0   
80992    14000773        103156371        CTNF  2015-12-23        0   
80993    14099483        103156415        CTFR  2015-12-21        0   
81003    13734066        103156527        CTFR  2015-12-23        0   
81022    14048997        103156808        CTFR  2015-12-21        0   
81028    14047445        103156831        CTFR  2015-12-21        0   
81031    14076998        103156841        CTFR  2015-12-23        0   
81038    14047253        103156861        CTFR  2015-12-23        0   
81039    14075409        103156885        CTNF  2015-12-23        0   
81043    14005794        103156928        CTNF  2015-12-23        0   
81065    14067599        103157213        CTNF  2015-12-23        0   
81071    14095198        103157300        CTFR  2015-12-23        0   
81073    14230432        103157309        CTNF  2015-12-23        0   
81076    14095744        103157354        CTNF  2015-12-23        0   
81079    14028901        103157359        CTNF  2015-12-23        0   
81080    14053111        103157363        CTFR  2015-12-23        0   
81099    14095613        103157558        CTFR  2015-12-23        0   
81108    13616496        103157761        CTNF  2015-12-23        0   
81118    14022146        103157958        CTFR  2015-12-23        0   
81128    14081088        103158109        CTFR  2015-12-23        0   
81138    14069392        103158276        CTNF  2015-12-23        0   
81156    14024983        103159236        CTNF  2015-12-23        0   
81196    14084593        103160361        CTNF  2015-12-23        0   
81198    14043983        103160368        CTFR  2015-12-23        0   
81207    14080196        103160683        CTNF  2015-12-23        0   
81223    14076640        103160930        CTNF  2015-12-23        0   
81251    14006504        103161774        CTFR  2015-12-23        0   
...           ...              ...         ...         ...      ...   
1627353  12761998  IXUPG7S5RXEAPX1        CTFR  2017-01-17        0   
1627549  13601976  IXUQ8C39RXEAPX3        CTNF  2017-01-17        0   
1627667  13442600  IXUQKQRXRXEAPX3        CTNF  2017-01-13        0   
1627925  14618897  IXUSAED2RXEAPX5        CTFR  2017-01-17        0   
1627943  13022616  IXVS8QVYRXEAPX5        CTNF  2017-01-18        0   
1627968  13086949  IXVSKBIERXEAPX5        CTFR  2017-01-18        0   
1628119  13068506  IXVTIZCMRXEAPX0        CTFR  2017-01-18        0   
1628195  13048136  IXVTVC8ARXEAPX1        CTFR  2017-01-18        0   
1628364  13061860  IXVUSCAFRXEAPX3        CTNF  2017-01-18        0   
1628424  13000622  IXVV0WG0RXEAPX5        CTNF  2017-01-18        0   
1628633  13063076  IXVVX1MTRXEAPX4        CTFR  2017-01-18        0   
1628664  13075849  IXVW2BA7RXEAPX4        CTFR  2017-01-18        0   
1628745  13018636  IXVWDNK5RXEAPX0        CTFR  2017-01-13        0   
1629568  13033130  IXVZW6GMRXEAPX5        CTFR  2017-01-18        0   
1629578  13089431  IXVZYXRZRXEAPX0        CTNF  2017-01-18        0   
1629880  13099240  IXW2VRZXRXEAPX1        CTFR  2017-01-18        0   
1629970  13078945  IXW4F3KGRXEAPX4        CTNF  2017-01-18        0   
1630167  13063828  IXW58TWPRXEAPX1        CTNF  2017-01-18        0   
1631542  12678648  IY1MS2J9RXEAPX3        CTNF  2017-01-19        0   
1631584  12705560  IY1MYG4DRXEAPX5        CTFR  2017-01-19        0   
1631705  12131022  IY1NGLJQRXEAPX1        CTNF  2017-01-17        0   
1631741  13011798  IY1NKLM2RXEAPX2        CTNF  2017-01-19        0   
1631950  13079074  IY1OC3U4RXEAPX4        CTNF  2017-01-17        0   
1632262  14548079  IY1PCAMFRXEAPX4        CTNF  2017-01-19        0   
1632264  12117422  IY1PCDU2RXEAPX

In [35]:
df[df['art_unit']==0]
#13423

app_id_x       ifw_number document_cd     mail_dt art_unit  \
1507378  13116344  IV9PRBGJRXEAPX0        CTNF  2016-11-10        0   
1507452  13082230  IV9QQOEIRXEAPX1        CTFR  2016-11-10        0   
1507921  13062592  IV9V4VOLRXEAPX5        CTFR  2016-11-10        0   
1507941  13162245  IV9V9C4CRXEAPX5        CTFR  2016-11-10        0   
1507964  13086602  IV9VEF9KRXEAPX0        CTFR  2016-11-10        0   
1508019  12021227  IV9WGHHGRXEAPX4        CTFR  2016-11-10        0   
1508033  13022182  IV9X8Q01RXEAPX4        CTFR  2016-11-10        0   
1508083  13037093  IV9XY3UDRXEAPX4        CTFR  2016-11-10        0   
1508158  13045035  IV9Z1IDERXEAPX0        CTFR  2016-11-10        0   
1508202  15239609  IV9ZMXE9RXEAPX0        CTNF  2016-11-09        0   
1508221  13024440  IVA0311TRXEAPX3        CTFR  2016-11-10        0   
1508291  13042641  IVAXHAHRRXEAPX5        CTFR  2016-11-14        0   
1508484  12068742  IVAYNV34RXEAPX1        CTFR  2016-11-14        0   
1508505  13166563  IVAYRIKSRXEAPX3        CTNF  2016-11-09        0   
1508511  13013461  IVAYS6B2RXEAPX4        CTNF  2016-11-14        0   
1508540  13102043  IVAYWV9SRXEAPX4        CTNF  2016-11-14        0   
1508653  13173390  IVAZIU4LRXEAPX4        CTFR  2016-11-14        0   
1508726  12096773  IVAZTHIGRXEAPX0        CTNF  2016-11-14        0   
1508971  13098725  IVB111Z6RXEAPX1        CTFR  2016-11-14        0   
1508998  13043665  IVB160X7RXEAPX3        CTNF  2016-11-14        0   
1509209  13066796  IVB29HD9RXEAPX0        CTNF  2016-11-09        0   
1509283  13002636  IVB2O7NHRXEAPX1        CTNF  2016-11-14        0   
1509308  13071757  IVB2T7A7RXEAPX0        CTNF  2016-11-14        0   
1509309  13080986  IVB2TIBXRXEAPX5        CTFR  2016-11-14        0   
1509787  13002980  IVB5YW0HRXEAPX2        CTNF  2016-11-14        0   
1510366  13033383  IVBBVYA8RXEAPX3        CTNF  2016-11-14        0   
1510389  13049429  IVBC3FNGRXEAPX2        CTFR  2016-11-14        0   
1510390  13073663  IVBC3Q7GRXEAPX4        CTFR  2016-11-14        0   
1510439  13011324  IVBCLWZPRXEAPX4        CTFR  2016-11-14        0   
1511986  12459581  IVCNB6X8RXEAPX2        CTFR  2016-11-10        0   
...           ...              ...         ...         ...      ...   
1778265  14310564  J4UQBHIPRXEAPX2        CTNF  2017-07-06        0   
1778274  14347016  J4UQDX99RXEAPX2        CTNF  2017-07-06        0   
1778276  14390384  J4UQDZ6IRXEAPX5        CTFR  2017-07-06        0   
1778287  14367964  J4UQG56FRXEAPX2        CTNF  2017-07-06        0   
1778319  12680586  J4UQR5X5RXEAPX1        CTFR  2017-07-06        0   
1778328  14391533  J4UQTEIIRXEAPX3        CTFR  2017-07-06        0   
1778332  12645179  J4UQU4LNRXEAPX4        CTFR  2017-07-06        0   
1778343  14318933  J4UQX3O2RXEAPX0        CTNF  2017-07-06        0   
1778348  14359812  J4UR0PQ7RXEAPX4        CTFR  2017-07-06        0   
1778350  14384345  J4UR14OZRXEAPX1        CTNF  2017-07-06        0   
1778357  14374729  J4UR4BKURXEAPX0        CTFR  2017-07-06        0   
1778359  14385226  J4UR4S2VRXEAPX3        CTNF  2017-07-06        0   
1778399  14340872  J4URIGT8RXEAPX1        CTNF  2017-07-06        0   
1778435  14307003  J4URTT97RXEAPX2        CTFR  2017-07-06        0   
1778496  12901709  J4USH16GRXEAPX3        CTNF  2017-07-06        0   
1778500  14397611  J4USK3N3RXEAPX2        CTFR  2017-07-06        0   
1778527  14310589  J4USX71ERXEAPX0        CTNF  2017-07-06        0   
1778540  14303788  J4UT9PF9RXEAPX5        CTFR  2017-07-06        0   
1778552  14332576  J4UTQ1F0RXEAPX1        CTFR  2017-07-06        0   
1778560  14370609  J4UTRM0NRXEAPX1        CTFR  2017-07-06        0   
1778578  12094939  J4UU3HEKRXEAPX5        CTFR  2017-07-06        0   
1778603  14396665  J4UUG6LARXEAPX4        CTNF  2017-07-06        0   
1778627  14380015  J4UUT6NSRXEAPX4        CTFR  2017-07-06        0   
1778629  14377889  J4UUW6FPRXEAPX4        CTNF  2017-07-06        0   
1778636  12416005  J4UV2Q5FRXEAPX

# 3. look at percentages of total applications compared to alice rejections in the class.
```
Percent of Alice rejections in USPC class 706: 2.3445902344590235
Percent of Alice rejections in USPC class 705: 35.66625497461232
```
Did not do pre-cleaning and filling in missing values where an art_unit is present but there is no uspc class

In [52]:
df[df['art_unit']=='2121'] #2173 rows
df[df['art_unit']==2121] #204 rows

app_id_x       ifw_number document_cd     mail_dt art_unit  \
1507899  14201952  IV9UYM8TRXEAPX2        CTNF  2016-11-10     2121   
1513909  13707921  IVIMVO3PRXEAPX4        CTFR  2016-11-16     2121   
1514127  14360242  IVJHHBH1RXEAPX1        CTNF  2016-11-17     2121   
1514987  13980476  IVJKH5GHRXEAPX2        CTFR  2016-11-17     2121   
1515115  13891560  IVJKX5KARXEAPX4        CTFR  2016-11-17     2121   
1518638  14088014  IVK2HK52RXEAPX3        CTNF  2016-11-17     2121   
1519349  14678720  IVKZTD5XRXEAPX5        CTFR  2016-11-16     2121   
1519467  14556212  IVL0962WRXEAPX3        CTNF  2016-11-18     2121   
1519698  14360109  IVL191JHRXEAPX0        CTNF  2016-11-18     2121   
1520367  13946857  IVL46MR0RXEAPX0        CTFR  2016-11-18     2121   
1524298  15273347  IVMEM9MYRXEAPX0        CTNF  2016-11-21     2121   
1524809  13837662  IVMHTY5CRXEAPX5        CTNF  2016-11-21     2121   
1524958  14450962  IVMJ5OEXRXEAPX1        CTNF  2016-11-21     2121   
1528289  14318682  IVS4WAIHRXEAPX2        CTNF  2016-11-23     2121   
1530331  13919438  IVTJBCTVRXEAPX0        CTFR  2016-11-25     2121   
1530358  12355894  IVTJENRPRXEAPX2        CTNF  2016-11-25     2121   
1530434  13708577  IVTJNUBRRXEAPX2        CTNF  2016-11-25     2121   
1531352  14862807  IVTMSZDHRXEAPX1        CTFR  2016-11-25     2121   
1531453  13996016  IVTN718IRXEAPX1        CTNF  2016-11-25     2121   
1533308  14481661  IVUY770VRXEAPX2        CTNF  2016-11-28     2121   
1533751  14342337  IVV0AKO8RXEAPX0        CTFR  2016-11-28     2121   
1535042  14028888  IVV7JNPIRXEAPX1        CTFR  2016-11-28     2121   
1536150  13827717  IVXVIV9FRXEAPX4        CTNF  2016-11-29     2121   
1536448  13753802  IVXY45UZRXEAPX5        CTFR  2016-11-29     2121   
1537016  14088819  IVZL1S65RXEAPX0        CTFR  2016-11-16     2121   
1537072  14001823  IVZL7HEBRXEAPX1        CTFR  2016-11-16     2121   
1537552  14066850  IW23Q0S4RXEAPX0        CTNF  2016-11-30     2121   
1537739  13957668  IW24IQLSRXEAPX3        CTNF  2016-11-30     2121   
1538025  14302208  IW25Z2XDRXEAPX0        CTNF  2016-11-30     2121   
1539118  13875213  IW2AEUWBRXEAPX4        CTNF  2016-11-30     2121   
...           ...              ...         ...         ...      ...   
1753461  14721926  J4A2SAWERXEAPX3        CTNF  2017-06-23     2121   
1754946  14483221  J4E5SA1ORXEAPX2        CTNF  2017-06-28     2121   
1754999  14481878  J4E6203QRXEAPX5        CTFR  2017-06-28     2121   
1755125  14631044  J4E6V76QRXEAPX0        CTNF  2017-06-28     2121   
1755242  14486597  J4E7VIU3RXEAPX4        CTFR  2017-06-28     2121   
1755380  14110279  J4E8TEIQRXEAPX5        CTNF  2017-06-28     2121   
1755409  14761991  J4E8YBYKRXEAPX5        CTFR  2017-06-28     2121   
1755636  14701539  J4EA9RXRRXEAPX4        CTNF  2017-06-28     2121   
1755815  14091795  J4EDA3TNRXEAPX2        CTFR  2017-06-28     2121   
1756013  14698476  J4EEW37JRXEAPX4        CTNF  2017-06-28     2121   
1756025  14079139  J4EEZCMMRXEAPX3        CTNF  2017-06-28     2121   
1756600  13946857  J4FIDM5URXEAPX4        CTNF  2017-06-29     2121   
1757137  13815710  J4FJWTECRXEAPX3        CTNF  2017-06-27     2121   
1757393  14712532  J4FKST15RXEAPX5        CTNF  2017-06-29     2121   
1759359  14826432  J4FS07BARXEAPX4        CTNF  2017-06-29     2121   
1760510  14678053  J4G026XERXEAPX1        CTNF  2017-06-28     2121   
1760887  14317244  J4G25HK4RXEAPX3        CTNF  2017-06-29     2121   
1761464  14725596  J4GYYDERRXEAPX3        CTNF  2017-06-30     2121   
1764248  14734297  J4HBNSXIRXEAPX2        CTNF  2017-06-30     2121   
1764620  14387337  J4HDUJO2RXEAPX5        CTNF  2017-06-30     2121   
1765889  13996016  J4HLJU2MRXEAPX3        CTFR  2017-06-30     2121   
1766368  14623297  J4IE62CYRXEAPX2        CTNF  2017-06-29     2121   
1766909  14309246  J4IG4Y30RXEAPX3        CTFR  2017-07-03     2121   
1771752  13282079  J4QZPV25RXEAPX2        CTFR  2017-07-07     2121   
1773524  14185884  J4SGHJBARXEAPX

In [56]:
print(len(df[df['uspc_class']=='706'])) #4649 rows
#df[df['uspc_class']==706] #0 rows
print(len(df[df['uspc_class']=='705']))

4649
72870


In [61]:
(len(df[(df['uspc_class']=='706')&(df['alice_in']==1)])/(len(df[df['uspc_class']=='706'])))*100

2.3445902344590235

In [67]:
print('Percent of Alice rejections in USPC class 706: {0}'.format((len(df[(df['uspc_class']=='706')&(df['alice_in']==1)])/(len(df[df['uspc_class']=='706'])))*100))
print('Percent of Alice rejections in USPC class 705: {0}'.format((len(df[(df['uspc_class']=='705')&(df['alice_in']==1)])/(len(df[df['uspc_class']=='705'])))*100))

Percent of Alice rejections in USPC class 706: 2.3445902344590235
Percent of Alice rejections in USPC class 705: 35.66625497461232


# 4. for 706 and 705: look at dataset for rows with art unit but no uspc class and fill in 705 or 706. My focus is on the uspc class not the art unit. 

In [75]:
print(len(df[df['uspc_class']=='0']))
print(len(df[df['uspc_class']==0]))
print(len(df[df['uspc_class'].isnull()]))
print(len(df[df['uspc_class'].isna()]))

0
0
0
0


# 5. look at percentages of total applications compared to alice rejections in the class.

# 6. create and export a csv file with only the 705 and 706 rejections